# Syntenizer3000
Tool for bacterial genome analysis including but not limited to synteny measurement.

Run build.sh in the folder with the source code to build an executable on a UNIX derived OS. GCC with at least C++11 support or equivalent must be installed.

Run without any input parameters to get a list of possible arguments and a short explanation of what they do.

Below is an example of running the program with a full suite of analysis enabled:

./Syntenizer --gffdir=/data/input --outdir=/data/output --genegroups=/data/input/gene_groups.csv --contigcolours=/data/input/ContigColours.csv --syntenizegroups --generatestraincharts --generatesyntenycharts --chromosomeidentifier=chromosome --plasmididentifier=Rh --fragmentidentifier=fragment --genegrouphighlightcolours=/data/input/GeneGroupHighlightColours.csv --ffndir=/data/input --generatefnagroups --presenceabsencematrix --gapitpresenceabsencematrix --generelationmatrix --classifyfragments --disambiguategroups

In the above example all input files, including *.gff, *.fnn gene groups and others are located in the folder '/data/input', and all files generated by the tool are output into the folder '/data/output'.

If the program has been configured to output circular strain chart data or gene group synteny map data then the next step is to render these using the provided R scripts 'GenerateCharts.R' and 'GenerateSyntenyMaps.R'. These scripts need to be rung with a single input parameter, which is the path to the folder with the chart or map data which in the above example would be '/data/output'.

Things to note:

If issues or uncertainties crop up while trying to use the program you can refer to the example dataset provided. A full suit of analysis can be performed using said dataset so it should serve as a guide on how input data should be formatted for the program to work.

Only GFF3 files generated by PROKKA have been tested. Other versions or files generated by other tools may contain variations that might cause incompatibility issues. In that case contact the author for possible solutions or fixes.

Two gene group file formats are supported. The direct gene group file output from ProteinOrtho, plus an internal format detailed below.
Each gene group in the internal format is defined on a ';' separated line starting with The group id as the first item. All subsequent items should be a contatanation of the strain id and gene id with the character '|' separating them. See gene_groups.csv in the example dataset for an example.

To generate basic circular charts the program only needs properly formatted GFF3 files and accompanying gene group file.
However, the GC3s lane requires the presence of gene sequence data. If these are part of the GFF3 files then they will be parsed automatically. Alternatively they can be provided using gene group *.fna files.

In order for contigs to be coloured according to their type (chromosome/plasmid/fragment) then the contig id's in the GFF3 files have to contain an string patter that identifies them as such. The program can then be configured to recognize these 

Contig colours...

OPTIONS
	--gffdir=dir
		Sets the directory where strain gff files are located. All files with the .gff extension will be loaded. The strain ID is           		set to the name of the file, with the the extension removed, i.e. the file calb.gff defines the genome of the calb strain.    		If contig size and contig sequence data are present in the gff file then these will be parsed. If they are not present the 
		size of each contig is set to the end of the final gene defined on said contig. Alternatively sequence data for each gene 		can be provided using .ffn files. Consult the example dataset for examples of file formatting.    
		If present additional information about genes will be parsed from .gff files such as gene function defined using the 					'product=' tag as seen in the example dataset.
    
		--outdir=dir
		Setz the output directory for all files generated by Syntenizer3000.

	--genegroups=file
		Set the file specifying gene groups. Syntenizer3000 can parse two different ways of encoding this information. The first 			is the direct output file format from the orthology tool ProteinOrtho. The second is an internal format, i.e. as below
    Group1: strain1|gene1 strain2|gene1
    Each tab delimited line starts with a group ID finished with a colon. Then a list of genes in said group each consisting                             
    of the ID of the gene strain and ID of the gene in said strain, separated by the | ascii character. 

	--ffndir=dir
		Sets the directory where gene sequence .ffn files are located. All files with the ffn extension are parsed, and the
    gene sequences found within are attached to their corresponding genes. Sequences for unknown genes are ignored.  

	--contigcolours=file
		Set the file specifying the contig colours.


	--syntenizegroups
		Generate synteny score for gene groups.

	--disambiguategroups
		Generate gene groups devoid of paralogs.

	--generatefnagroups
		Generate gene group fna files.

	--generatestraincharts
		Generate data files for strain chart rendering.

	--generatesyntenycharts
		Generate data files for group synteny chart rendering.

	--presenceabsencematrix
		Generage presence/absence matrix based on gene groups.

	--gapitpresenceabsencematrix
		Generage GAPIT formatted presence/absence matrix based on gene groups.

	--generelationmatrix
		Generage gene relation matrix based on gene groups.

	--chromosomeidentifier=pattern
		Set the contig name character pattern that identifies a chromosomal contig.

	--plasmididentifier=pattern
		Set the contig name character pattern that identifies a plasmid contig.

	--fragmentidentifier=pattern
		Set the contig name character pattern that identifies an unclassified contig.

	--syntenizegenepairs=file
		Set the file specifying gene pairs to be syntenized.

	--genegrouphighlightcolours=file
		Set the file specifying gene group highlight colours on strain charts.

	--classifyfragments
		Generate fragment classifications based on synteny and gene group information.


To Be Completed
